### PR TITLE
Use the exec-sidecar as a healthz probe for the syndns container

### PR DESCRIPTION
**Pull request by bprashanth**
Gives kube-dns a liveness probe that uses the healthz-sidecar container. Also given etcd a emptyDir volume so cluster dns survives across etcd restarts.

Related: #10659, #10926, #10925
@dchen1107 @thockin

### CODE REVIEW

File: cluster/addons/dns/skydns-rc.yaml.in [hunk 1]
```diff
@@ -61,4 +66,26 @@ spec:
...
+        livenessProbe:
+          httpGet:
+            path: /healthz
+            port: 8080
+            scheme: HTTP
+          initialDelaySeconds: 30
+          timeoutSeconds: 5
+      - name: healthz
+        image: gcr.io/google_containers/healthz-server:1.0
+        resources:
```
Comment (bprashanth) [inline cluster/addons/dns/skydns-rc.yaml.in]: I think this container is tiny enough that it will fit in somewhere. Do we need to give it a tenth of a core?

Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: We don't have any "outer" shared limits yet that would let you get away with 0.  You could spec 10m for CPU :)

**Comment by thockin** [comment]
Interestingly, we could use this same health-check on every container in the pod now, and get more reliable restarts (in truth, failure to resolve is more likely to be a kube2sky bug than skydns).  Maybe after 1.0.  #8194

File: cluster/addons/dns/skydns-rc.yaml.in [hunk 2]
```diff
@@ -28,12 +28,17 @@ spec:
             memory: 50Mi
         command:
         - /usr/local/bin/etcd
+        - -data-dir
```
Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: This is unrelated to this PR ?

File: cluster/addons/dns/skydns-rc.yaml.in [hunk 3]
```diff
@@ -61,4 +66,26 @@ spec:
...
+            port: 8080
+            scheme: HTTP
+          initialDelaySeconds: 30
+          timeoutSeconds: 5
+      - name: healthz
+        image: gcr.io/google_containers/healthz-server:1.0
+        resources:
+          limits:
+            cpu: 100m
+            memory: 5Mi
```
Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: is 5 MB enough for a go program and exec()s ?

**Comment by thockin** [comment]
LGTM but you have the data dir PR mixed in.

File: cluster/addons/dns/skydns-rc.yaml.in (hunk 1 above)
Comment (bprashanth) [inline cluster/addons/dns/skydns-rc.yaml.in]: I was planning on 10m. What do you mean by outer shared limits? like a limit range on the ns (that kube-system doesn't have).

File: cluster/addons/dns/skydns-rc.yaml.in (hunk 3 above)
Comment (bprashanth) [inline cluster/addons/dns/skydns-rc.yaml.in]: Yeah it seemed fine for this exec (what I saw was actually kb). Can make it 10m.

File: cluster/addons/dns/skydns-rc.yaml.in (hunk 2 above)
Comment (bprashanth) [inline cluster/addons/dns/skydns-rc.yaml.in]: will split it out, makes sense because we might revert one. testing this was hard without data dir.

Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: yeah, needs to be 2 PRs

File: cluster/addons/dns/skydns-rc.yaml.in (hunk 1 above)
Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: 10m should be fine, you have 100m now.

"outer" means a limit for the pod, such that something with 0 CPU could use whatever CPU was free in the pod

File: cluster/addons/dns/skydns-rc.yaml.in (hunk 3 above)
Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: Really?  I have never seen a Go program less than a few MB.  I'm skittish about OOM for Go + GC...

Comment (bprashanth) [inline cluster/addons/dns/skydns-rc.yaml.in]: better safe than sorry

**Comment by bprashanth** [comment]
this is blocked on me updating the image once https://github.com/GoogleCloudPlatform/kubernetes/pull/10925 goes in

**Comment by thockin** [comment]
LGTM

Jenkins failed

**Comment by thockin** [comment]
@ArtfulCoder for 2nd eyes

**Comment by bprashanth** [comment]
Jenkins failed because i haven't updated the image yet, will do and rerun

**Comment by bprashanth** [comment]
@k8s-bot ok to test

File: cluster/addons/dns/skydns-rc.yaml.in (hunk 3 above)
Comment (dchen1107) [inline cluster/addons/dns/skydns-rc.yaml.in]: 4MB is minimal you can specified, otherwise docker would reject it. But I am not comfortable for less than 10M. :-)

Comment (bprashanth) [inline cluster/addons/dns/skydns-rc.yaml.in]: Yeah, totally makes sense especially given Tim's gc point. It's currently set to 20megs.

**Comment by bprashanth** [comment]
@dchen1107 for second lgtm and @rjnagal for merge

**Comment by thockin** [comment]
one of you or @ArtfulCoder will need to resolve v6/v7/v8 :)

**Comment by thockin** [comment]
@jnagal for priority merge

**Comment by jnagal** [comment]
waiting on @dchen1107 for the second nod.

@thockin: use @rjnagal to get my immediate attention. The other one is one hop away - goes to my brother who forwards it back to me.

**Comment by thockin** [comment]
dammit, sorry - I know this and I keep screwing it up :)

**Comment by thockin** [comment]
if I send him enough email will he give it back to you? :)

File: cluster/addons/dns/skydns-rc.yaml.in [hunk 4]
```diff
@@ -61,4 +61,23 @@ spec:
...
+          initialDelaySeconds: 30
+          timeoutSeconds: 5
+      - name: healthz
+        image: gcr.io/google_containers/exechealthz:1.0
+        resources:
+          limits:
+            cpu: 10m
+            memory: 20Mi
+        args:
+        - -cmd=nslookup kubernetes.default.svc.{{ pillar['dns_domain'] }} localhost >/dev/null
```
Comment (ArtfulCoder) [inline cluster/addons/dns/skydns-rc.yaml.in]: you should not need to say pillar['dns_domain']
just kubernetes.default.svc should be enough

Comment (ArtfulCoder) [inline cluster/addons/dns/skydns-rc.yaml.in]: Never mind my prev comment.. 
we need the whole path..

Comment (thockin) [inline cluster/addons/dns/skydns-rc.yaml.in]: I want this to be an absoluet lookup

Comment (ArtfulCoder) [inline cluster/addons/dns/skydns-rc.yaml.in]: relative path wont work either..
we dont have the search domain suffixes on this pod.

**Comment by thockin** [comment]
Anything holding this up?  I want to get it in ASAP

**Comment by thockin** [comment]
@ArtfulCoder mark it as OK to merge if you're happy with it.  I don't care who has to rebase, I want these two PRs in - pronto

**Comment by rjnagal** [comment]
thanks, merging.

**Comment by dchen1107** [comment]
LGTM
